### PR TITLE
Fix memory leak in ReactChildrenMutationWarningHook for SSR

### LINKED ISSUE #7406: Memory leak in React 15.3.0 non-production server side rendering
A memory leak appears to have been introduced to server side rendering in React 15.3.0 when `NODE_ENV != production`.

With `NODE_ENV=development`, the size of the heap grows proportionally to the number of requests received. This was confirmed in both Node v6.3.0 and v0.12.3. Also it was **confirmed to work correctly in React 15.2.0**.

_To reproduce_: Start the app configured for server-side rendering with `NODE_ENV=development` (or just not production). Inspect the V8 heap before and after hitting your server with a number of requests. I used [heapdump](https://github.com/bnoordhuis/node-heapdump) to take the snapshots and [wrk](https://github.com/wg/wrk) to hit my app with a bunch of requests. Load the snapshots into chrome dev tools profiler. Here is what I see:
screen shot 2016-08-02 at 12 47 34 pm

The culprit appears to be the [react children mutation warning](https://github.com/facebook/react/blob/328fc75bc910032ff37dc9d93429fb14b476bd0d/src/renderers/shared/hooks/ReactChildrenMutationWarningHook.js#L18). `onBeforeComponentMount` saves a reference to a given React Element and all of its props in the `elements` object, but the cleanup steps `onComponentHasMounted` or `onComponentHasUpdated` are never called on the server and thus the `elements` object is ever growing.

**Which versions of React, and which browser / OS are affected by this issue? Did this work in previous versions of React?**
react - 15.3.0
react-dom - 15.3.0
node - 6.3.0 || 0.12.3
express - ^4.0.0

### PR DESCRIPTION
**Pull request by keyanzhang**
Fixes #7406. Thanks @franjohn21 for providing such a detailed investigation!

Reviewers: @gaearon @jimfb

### CODE REVIEW

**Comment by lynn** [comment]
Awesome job! Will there be a patch release for this fix?

File: src/renderers/dom/server/__tests__/ReactServerRendering-test.js [hunk 1]
```diff
@@ -538,4 +538,38 @@ describe('ReactServerRendering', function() {
...
+    var element = <div>{children}</div>;
+    children[1] = <p key={1} />; // Mutation is illegal
+    ReactServerRendering.renderToString(element);
+    expect(console.error.calls.count()).toBe(1);
+    expect(normalizeCodeLocInfo(console.error.calls.argsFor(0)[0])).toBe(
+      'Warning: Component\'s children should not be mutated.\n    in div (at **)'
+    );
+  });
+
+  it('should warn when children are mutated', function() {
```
Comment (gaearon) [inline src/renderers/dom/server/__tests__/ReactServerRendering-test.js]: Nit: not very clear why there are two tests, and how they are different (right not the name only differences is starting with "should")

**Comment by gaearon** [comment]
This will probably go in 15.3.1.

@franjohn21 Can you build from master and verify this fixes it?

**Comment by gaearon** [comment]
This looks right but I'd appreciate if you could verify this indeed fixes the leak.

**Comment by franjohn21** [comment]
@keyanzhang @gaearon I can confirm I am seeing the memory leak removed with this change. Thanks for the quick turnaround time!

**Comment by keyanzhang** [comment]
Note to self: #7424 and #7426

**Comment by gaearon** [comment]
I’m going to get this in because it’s better than status quo and I also need it for __DEV__ perf regression I’m tackling right now.

File: src/renderers/shared/hooks/ReactChildrenMutationWarningHook.js [hunk 2]
```diff
@@ -46,21 +44,13 @@ function handleElement(debugID, element) {
...
-var ReactDOMUnknownPropertyHook = {
-  onBeforeMountComponent(debugID, element) {
-    elements[debugID] = element;
-  },
-  onBeforeUpdateComponent(debugID, element) {
-    elements[debugID] = element;
-  },
-  onComponentHasMounted(debugID) {
-    handleElement(debugID, elements[debugID]);
-    delete elements[debugID];
```
Comment (gaearon) [inline src/renderers/shared/hooks/ReactChildrenMutationWarningHook.js]: For future reference: `delete` is super slow on V8 with many items when using numeric keys. Even `toString()` is not enough—you have to prepend them with something like `.`.

**Pull request merged**